mutable_buffer/src/table.rs

```diff
@@ -237,6 +240,106 @@ impl Table {
+                            if !insert_column.is_field() {
+                                InternalColumnTypeMismatch {
+                                    column_id,
+                                    expected_column_type: c.type_description(),
+                                    actual_column_type: values.type_description(),
+                                }
+                                .fail()?
+                            };
+                        }
+                        (entry::TypedValuesIterator::String(_), Column::Tag(_, _)) => {
+                            if !insert_column.is_tag() {
+                                InternalColumnTypeMismatch {
+                                    column_id,
+                                    expected_column_type: c.type_description(),
+                                    actual_column_type: values.type_description(),
+                                }
+                                .fail()?
+                            };
+                        }
+                        _ => InternalColumnTypeMismatch {
+                            column_id,
+                            expected_column_type: c.type_description(),
+                            actual_column_type: values.type_description(),
+                        }
+                        .fail()?,
+                    }
+                }
+
+                Ok((column_id, insert_column.logical_type(), values))
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        let row_count_before_insert = self.row_count();
+
+        for (column_id, logical_type, values) in columns_with_inserts.into_iter() {
+            match self.columns.get_mut(&column_id) {
+                Some(c) => c
+                    .push_typed_values(dictionary, logical_type, values)
+                    .with_context(|| {
+                        let column = dictionary.lookup_id(column_id).unwrap_or("unknown");
+                        ColumnError { column }
+                    })?,
+                None => {
+                    self.columns.insert(
+                        column_id,
+                        Column::new_from_typed_values(
+                            dictionary,
+                            row_count_before_insert,
+                            logical_type,
+                            values,
+                        ),
+                    );
+                }
+            }
+        }
+
+        // ensure all columns have the same number of rows as the one with the most.
+        // This adds nulls to the columns that weren't included in this write
+        let max_row_count = self
+            .columns
+            .values()
+            .fold(row_count_before_insert, |max, col| {
+                let len = col.len();
+                if max < len {
+                    len
+                } else {
+                    max
+                }
+            });
```

Comment:
Hm but why not:

```suggestion
        let max_row_count = self
            .columns
            .values()
            .map(|col| col.len())
            .max()
            .expect("columns isn't empty");
```

? 😆

Reply:
I wanted to suggest this too but was too lazy to check whether a `max(row_count_before_insert, max_row_count)` was needed too; and at the end of day the whole expression was way longer than just using fold with a seed

Reply:
Yeah it seems like `row_count_before_insert` should be less than or equal to the max as I calculated here, but I don't know enough to be 100% confident.